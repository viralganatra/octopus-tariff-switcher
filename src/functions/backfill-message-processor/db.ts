import {
  BatchWriteItemCommand,
  DynamoDBClient,
  type AttributeValue,
} from '@aws-sdk/client-dynamodb';
import { marshall } from '@aws-sdk/util-dynamodb';
import { Resource } from 'sst';
import { format, parseISO } from 'date-fns';
import {
  getDailyUsageCostByTariff,
  getUnitRatesWithCost,
} from '../tariff-switcher/cost-calculator';
import type { TariffData } from '../backfill-message-publisher/api-data';
import { getDateInUTC } from '../../utils/helpers';
import { batchWithRetry } from '../../utils/fetch';
import type { DailyUsage } from './schema';

const BATCH_SIZE = 25;
const client = new DynamoDBClient();

function calculateCost(item: TariffData) {
  return getDailyUsageCostByTariff({
    standingCharge: item.standingCharge,
    consumptionUnitRates: item.consumption,
    tariffUnitRates: item.unitRates,
  });
}

function createWriteRequestsForDay(dailyUsage: DailyUsage) {
  const writeRequests: { PutRequest: { Item: Record<string, AttributeValue> } }[] = [];

  // Individual halfâ€‘hour reads
  const unitRatesWithCost = getUnitRatesWithCost({
    tariffUnitRates: dailyUsage.unitRates,
    consumptionUnitRates: dailyUsage.consumption,
  });

  for (const unitRate of unitRatesWithCost) {
    const readAtUTC = getDateInUTC(parseISO(unitRate.readAt));
    const timeKey = format(readAtUTC, 'HH:mm');

    writeRequests.push({
      PutRequest: {
        Item: marshall({
          PK: `DATE#${dailyUsage.isoDate}`,
          SK: `USAGE#${timeKey}`,
          readAt: unitRate.readAt,
          consumptionDelta: unitRate.consumptionDelta,
          costDelta: unitRate.unitCostInPence,
        }),
      },
    });
  }

  // Daily tariff cost
  const dailyTariffCost = calculateCost(dailyUsage);

  writeRequests.push({
    PutRequest: {
      Item: marshall({
        PK: `DATE#${dailyUsage.isoDate}`,
        SK: `TARIFF#${dailyUsage.id}`,
        tariffName: dailyUsage.tariffName,
        tariffId: dailyUsage.id,
        tariffCost: dailyTariffCost,
        tariffStandingCharge: dailyUsage.standingCharge,
      }),
    },
  });

  // Daily consumption total
  let dailyKwhTotal = 0;

  for (const unitConsumption of dailyUsage.consumption) {
    dailyKwhTotal += unitConsumption.consumptionDelta;
  }

  writeRequests.push({
    PutRequest: {
      Item: marshall({
        PK: `DATE#${dailyUsage.isoDate}`,
        SK: 'TOTAL',
        dailyKwhTotal,
      }),
    },
  });

  return writeRequests;
}

async function batchWriteWithRetry(writeRequests: ReturnType<typeof createWriteRequestsForDay>) {
  return batchWithRetry({
    entries: writeRequests,
    batchSize: BATCH_SIZE,
    sendBatch: async (batch) => {
      const command = new BatchWriteItemCommand({
        RequestItems: {
          [Resource.OctopusTariffSwitcherDailyUsageTable.name]: batch,
        },
      });

      const response = await client.send(command);

      const unprocessedItems =
        response.UnprocessedItems?.[Resource.OctopusTariffSwitcherDailyUsageTable.name] ?? [];

      const failedItems = unprocessedItems.filter(
        (item): item is { PutRequest: { Item: Record<string, AttributeValue> } } =>
          item.PutRequest !== undefined,
      );

      const failedKeys = failedItems
        .map((item) => {
          const pk = item.PutRequest.Item.PK?.S ?? 'unknown pk';
          const sk = item.PutRequest.Item.SK?.S ?? 'unknown sk';

          return `${pk}::${sk}`;
        })
        .join(', ');

      return {
        failed: failedItems,
        reason: `Failed to batch write to DynamoDB: ${failedKeys}`,
      };
    },
  });
}

export function saveDailyUsage(dailyUsage: DailyUsage) {
  const writeRequests = createWriteRequestsForDay(dailyUsage);

  return batchWriteWithRetry(writeRequests);
}
