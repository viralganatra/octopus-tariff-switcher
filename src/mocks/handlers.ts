import { http, HttpResponse, graphql } from 'msw';
import {
  accountFixture,
  productAgileFixture,
  productCosyFixture,
  productGoFixture,
  productsFixture,
  standingChargeFixture,
  telemetryFixture,
  unitRatesFixture,
} from './fixtures';

export const handlers = [
  http.get('https://api.octopus.energy/v1/products', () => {
    return HttpResponse.json(productsFixture);
  }),
  http.get('https://api.octopus.energy/v1/products/AGILE-24-10-01/', () => {
    return HttpResponse.json(productAgileFixture);
  }),
  http.get('https://api.octopus.energy/v1/products/COSY-22-12-08/', () => {
    return HttpResponse.json(productCosyFixture);
  }),
  http.get('https://api.octopus.energy/v1/products/GO-VAR-22-10-14/', () => {
    return HttpResponse.json(productGoFixture);
  }),
  http.get(
    'https://api.octopus.energy/v1/products/:tariffCode/electricity-tariffs/:productCode/standing-charges/',
    () => {
      return HttpResponse.json(standingChargeFixture);
    },
  ),
  http.get(
    'https://api.octopus.energy/v1/products/:tariffCode/electricity-tariffs/:productCode/standard-unit-rates',
    () => {
      return HttpResponse.json(unitRatesFixture);
    },
  ),
  graphql.mutation('ObtainKrakenToken', () => {
    return HttpResponse.json({
      data: {
        obtainKrakenToken: {
          token: 'foo',
        },
      },
    });
  }),
  graphql.query('Account', () => {
    return HttpResponse.json({
      data: accountFixture,
    });
  }),
  graphql.query('smartMeterTelemetry', () => {
    return HttpResponse.json({
      data: telemetryFixture,
    });
  }),
];
