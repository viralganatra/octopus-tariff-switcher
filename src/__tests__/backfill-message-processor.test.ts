import type { Context } from 'aws-lambda';
import { BatchWriteItemCommand, DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { mockClient } from 'aws-sdk-client-mock';
import { messagesFromQueueFixture } from '../mocks/fixtures';
import { processBackfillQueue } from '../backfill-message-processor';

const dynamoDbMock = mockClient(DynamoDBClient);

describe('Backfill Message Processor', () => {
  const context = {} as Context;

  beforeEach(() => {
    vi.setSystemTime(new Date(2025, 2, 3));
    dynamoDbMock.reset();
  });

  it('should batch write messages successfully', async () => {
    dynamoDbMock.on(BatchWriteItemCommand).resolves({ UnprocessedItems: {} });

    const promise = processBackfillQueue({ Records: messagesFromQueueFixture }, context);

    await vi.runAllTimersAsync();
    const data = await promise;

    expect(data).toMatchObject({ batchItemFailures: [] });
    expect(dynamoDbMock).toHaveReceivedCommandTimes(BatchWriteItemCommand, 6);

    const sentCommand1 = dynamoDbMock.commandCalls(BatchWriteItemCommand)[0]!.args[0].input;
    const items1 = sentCommand1.RequestItems!['test-table'] ?? [];

    expect(sentCommand1.RequestItems).toHaveProperty('test-table');
    expect(Object.values(items1)).toHaveLength(25);
    expect(items1).toMatchSnapshot();

    const sentCommand2 = dynamoDbMock.commandCalls(BatchWriteItemCommand)[1]!.args[0].input;
    const items2 = sentCommand2.RequestItems!['test-table'] ?? [];
    expect(items2).toMatchSnapshot();
  });

  it('should retry unprocessed items', async () => {
    dynamoDbMock
      .on(BatchWriteItemCommand)
      .resolvesOnce({
        UnprocessedItems: {
          'test-table': [
            {
              PutRequest: {
                Item: {
                  PK: {
                    S: 'DATE#2025-04-26',
                  },
                  SK: {
                    S: 'USAGE#00:30',
                  },
                  consumptionDelta: {
                    N: '85',
                  },
                  costDelta: {
                    N: '1.3325',
                  },
                  readAt: {
                    S: '2025-04-26T01:30:00+01:00',
                  },
                },
              },
            },
          ],
        },
      })
      .resolves({
        UnprocessedItems: {},
      });

    const promise = processBackfillQueue({ Records: messagesFromQueueFixture }, context);

    await vi.runAllTimersAsync();
    const data = await promise;

    expect(data).toMatchObject({ batchItemFailures: [] });
    expect(dynamoDbMock).toHaveReceivedCommandTimes(BatchWriteItemCommand, 7);
  });

  it('should fail gracefully if any retries fail', async () => {
    dynamoDbMock.on(BatchWriteItemCommand).resolves({
      UnprocessedItems: {
        'test-table': [
          {
            PutRequest: {
              Item: {
                PK: {
                  S: 'DATE#2025-04-26',
                },
                SK: {
                  S: 'USAGE#00:30',
                },
                consumptionDelta: {
                  N: '85',
                },
                costDelta: {
                  N: '1.3325',
                },
                readAt: {
                  S: '2025-04-26T01:30:00+01:00',
                },
              },
            },
          },
        ],
      },
    });

    const promise = processBackfillQueue({ Records: messagesFromQueueFixture }, context);

    await vi.runAllTimersAsync();
    const data = await promise;

    expect(data).toMatchObject({
      batchItemFailures: [
        {
          itemIdentifier: 'b21cf427-381b-4067-9626-baf3ebc1c2ea',
        },
        {
          itemIdentifier: '91f7053c-8de8-4da5-b640-09ce841c7bd5',
        },
        {
          itemIdentifier: '75aa7e13-1155-4389-9d08-b5966e5b870e',
        },
      ],
    });
    expect(dynamoDbMock).toHaveReceivedCommandTimes(BatchWriteItemCommand, 12);
  });
});
