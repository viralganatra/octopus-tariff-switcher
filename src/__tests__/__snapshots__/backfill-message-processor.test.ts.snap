// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Backfill Message Processor > should batch write messages successfully 1`] = `
[
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#00:00",
        },
        "consumptionDelta": {
          "N": "452",
        },
        "costDelta": {
          "N": "7.0668",
        },
        "readAt": {
          "S": "2025-04-26T01:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#00:30",
        },
        "consumptionDelta": {
          "N": "85",
        },
        "costDelta": {
          "N": "1.3325",
        },
        "readAt": {
          "S": "2025-04-26T01:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#01:00",
        },
        "consumptionDelta": {
          "N": "93",
        },
        "costDelta": {
          "N": "1.5028",
        },
        "readAt": {
          "S": "2025-04-26T02:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#01:30",
        },
        "consumptionDelta": {
          "N": "107",
        },
        "costDelta": {
          "N": "1.6448",
        },
        "readAt": {
          "S": "2025-04-26T02:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#02:00",
        },
        "consumptionDelta": {
          "N": "81",
        },
        "costDelta": {
          "N": "1.2868",
        },
        "readAt": {
          "S": "2025-04-26T03:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#02:30",
        },
        "consumptionDelta": {
          "N": "63",
        },
        "costDelta": {
          "N": "0.9678",
        },
        "readAt": {
          "S": "2025-04-26T03:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#03:00",
        },
        "consumptionDelta": {
          "N": "83",
        },
        "costDelta": {
          "N": "1.3543",
        },
        "readAt": {
          "S": "2025-04-26T04:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#03:30",
        },
        "consumptionDelta": {
          "N": "58",
        },
        "costDelta": {
          "N": "0.9452",
        },
        "readAt": {
          "S": "2025-04-26T04:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#04:00",
        },
        "consumptionDelta": {
          "N": "77",
        },
        "costDelta": {
          "N": "1.2524",
        },
        "readAt": {
          "S": "2025-04-26T05:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#04:30",
        },
        "consumptionDelta": {
          "N": "64",
        },
        "costDelta": {
          "N": "1.0134",
        },
        "readAt": {
          "S": "2025-04-26T05:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#05:00",
        },
        "consumptionDelta": {
          "N": "71",
        },
        "costDelta": {
          "N": "1.1898",
        },
        "readAt": {
          "S": "2025-04-26T06:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#05:30",
        },
        "consumptionDelta": {
          "N": "87",
        },
        "costDelta": {
          "N": "1.5118",
        },
        "readAt": {
          "S": "2025-04-26T06:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#06:00",
        },
        "consumptionDelta": {
          "N": "135",
        },
        "costDelta": {
          "N": "2.3814",
        },
        "readAt": {
          "S": "2025-04-26T07:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#06:30",
        },
        "consumptionDelta": {
          "N": "88",
        },
        "costDelta": {
          "N": "1.5717",
        },
        "readAt": {
          "S": "2025-04-26T07:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#07:00",
        },
        "consumptionDelta": {
          "N": "59",
        },
        "costDelta": {
          "N": "1.1362",
        },
        "readAt": {
          "S": "2025-04-26T08:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#07:30",
        },
        "consumptionDelta": {
          "N": "70",
        },
        "costDelta": {
          "N": "1.3252",
        },
        "readAt": {
          "S": "2025-04-26T08:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#08:00",
        },
        "consumptionDelta": {
          "N": "61",
        },
        "costDelta": {
          "N": "1.2253",
        },
        "readAt": {
          "S": "2025-04-26T09:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#08:30",
        },
        "consumptionDelta": {
          "N": "69",
        },
        "costDelta": {
          "N": "1.2976",
        },
        "readAt": {
          "S": "2025-04-26T09:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#09:00",
        },
        "consumptionDelta": {
          "N": "63",
        },
        "costDelta": {
          "N": "1.2866",
        },
        "readAt": {
          "S": "2025-04-26T10:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#09:30",
        },
        "consumptionDelta": {
          "N": "108",
        },
        "costDelta": {
          "N": "2.124",
        },
        "readAt": {
          "S": "2025-04-26T10:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#10:00",
        },
        "consumptionDelta": {
          "N": "317",
        },
        "costDelta": {
          "N": "5.6052",
        },
        "readAt": {
          "S": "2025-04-26T11:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#10:30",
        },
        "consumptionDelta": {
          "N": "97",
        },
        "costDelta": {
          "N": "1.5827",
        },
        "readAt": {
          "S": "2025-04-26T11:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#11:00",
        },
        "consumptionDelta": {
          "N": "81",
        },
        "costDelta": {
          "N": "1.3055",
        },
        "readAt": {
          "S": "2025-04-26T12:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#11:30",
        },
        "consumptionDelta": {
          "N": "129",
        },
        "costDelta": {
          "N": "2.0155",
        },
        "readAt": {
          "S": "2025-04-26T12:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#12:00",
        },
        "consumptionDelta": {
          "N": "113",
        },
        "costDelta": {
          "N": "1.7442",
        },
        "readAt": {
          "S": "2025-04-26T13:00:00+01:00",
        },
      },
    },
  },
]
`;

exports[`Backfill Message Processor > should batch write messages successfully 2`] = `
[
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#12:30",
        },
        "consumptionDelta": {
          "N": "110",
        },
        "costDelta": {
          "N": "1.6332",
        },
        "readAt": {
          "S": "2025-04-26T13:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#13:00",
        },
        "consumptionDelta": {
          "N": "103",
        },
        "costDelta": {
          "N": "1.5909",
        },
        "readAt": {
          "S": "2025-04-26T14:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#13:30",
        },
        "consumptionDelta": {
          "N": "539",
        },
        "costDelta": {
          "N": "8.3195",
        },
        "readAt": {
          "S": "2025-04-26T14:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#14:00",
        },
        "consumptionDelta": {
          "N": "159",
        },
        "costDelta": {
          "N": "2.4191",
        },
        "readAt": {
          "S": "2025-04-26T15:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#14:30",
        },
        "consumptionDelta": {
          "N": "104",
        },
        "costDelta": {
          "N": "1.6304",
        },
        "readAt": {
          "S": "2025-04-26T15:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#15:00",
        },
        "consumptionDelta": {
          "N": "85",
        },
        "costDelta": {
          "N": "2.3794",
        },
        "readAt": {
          "S": "2025-04-26T16:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#15:30",
        },
        "consumptionDelta": {
          "N": "96",
        },
        "costDelta": {
          "N": "2.8073",
        },
        "readAt": {
          "S": "2025-04-26T16:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#16:00",
        },
        "consumptionDelta": {
          "N": "810",
        },
        "costDelta": {
          "N": "23.5589",
        },
        "readAt": {
          "S": "2025-04-26T17:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#16:30",
        },
        "consumptionDelta": {
          "N": "692",
        },
        "costDelta": {
          "N": "21.58",
        },
        "readAt": {
          "S": "2025-04-26T17:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#17:00",
        },
        "consumptionDelta": {
          "N": "98",
        },
        "costDelta": {
          "N": "2.9903",
        },
        "readAt": {
          "S": "2025-04-26T18:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#17:30",
        },
        "consumptionDelta": {
          "N": "133",
        },
        "costDelta": {
          "N": "4.3375",
        },
        "readAt": {
          "S": "2025-04-26T18:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#18:00",
        },
        "consumptionDelta": {
          "N": "99",
        },
        "costDelta": {
          "N": "2.1434",
        },
        "readAt": {
          "S": "2025-04-26T19:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#18:30",
        },
        "consumptionDelta": {
          "N": "122",
        },
        "costDelta": {
          "N": "2.6465",
        },
        "readAt": {
          "S": "2025-04-26T19:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#19:00",
        },
        "consumptionDelta": {
          "N": "124",
        },
        "costDelta": {
          "N": "2.6053",
        },
        "readAt": {
          "S": "2025-04-26T20:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#19:30",
        },
        "consumptionDelta": {
          "N": "213",
        },
        "costDelta": {
          "N": "4.5558",
        },
        "readAt": {
          "S": "2025-04-26T20:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#20:00",
        },
        "consumptionDelta": {
          "N": "183",
        },
        "costDelta": {
          "N": "3.6028",
        },
        "readAt": {
          "S": "2025-04-26T21:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#20:30",
        },
        "consumptionDelta": {
          "N": "166",
        },
        "costDelta": {
          "N": "2.9213",
        },
        "readAt": {
          "S": "2025-04-26T21:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#21:00",
        },
        "consumptionDelta": {
          "N": "136",
        },
        "costDelta": {
          "N": "2.3148",
        },
        "readAt": {
          "S": "2025-04-26T22:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#21:30",
        },
        "consumptionDelta": {
          "N": "129",
        },
        "costDelta": {
          "N": "2.1144",
        },
        "readAt": {
          "S": "2025-04-26T22:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#22:00",
        },
        "consumptionDelta": {
          "N": "89",
        },
        "costDelta": {
          "N": "1.4915",
        },
        "readAt": {
          "S": "2025-04-26T23:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#22:30",
        },
        "consumptionDelta": {
          "N": "71",
        },
        "costDelta": {
          "N": "1.1824",
        },
        "readAt": {
          "S": "2025-04-26T23:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#23:00",
        },
        "consumptionDelta": {
          "N": "80",
        },
        "costDelta": {
          "N": "1.323",
        },
        "readAt": {
          "S": "2025-04-27T00:00:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "USAGE#23:30",
        },
        "consumptionDelta": {
          "N": "81",
        },
        "costDelta": {
          "N": "1.411",
        },
        "readAt": {
          "S": "2025-04-27T00:30:00+01:00",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "TARIFF#agile",
        },
        "tariffCost": {
          "N": "196.0163",
        },
        "tariffId": {
          "S": "agile",
        },
        "tariffName": {
          "S": "Agile Octopus",
        },
        "tariffStandingCharge": {
          "N": "48.7881",
        },
      },
    },
  },
  {
    "PutRequest": {
      "Item": {
        "PK": {
          "S": "DATE#2025-04-26",
        },
        "SK": {
          "S": "TOTAL",
        },
        "dailyKwhTotal": {
          "N": "7133",
        },
      },
    },
  },
]
`;
